darts/models/transformer_model.py

```diff
@@ -0,0 +1,477 @@
+        div_term = torch.exp(torch.arange(0, d_model, 2).float() * (-math.log(10000.0) / d_model))
+        pe[:, 0::2] = torch.sin(position * div_term)
+        pe[:, 1::2] = torch.cos(position * div_term)
+        pe = pe.unsqueeze(0).transpose(0, 1)
+        self.register_buffer('pe', pe)
+
+    def forward(self, x):
+        x = x + self.pe[:x.size(0), :]
+        return self.dropout(x)
+
+
+class _TransformerModule(nn.Module):
+    def __init__(self,
+                 input_size: int,
+                 output_length: int,
+                 output_size: int,
+                 d_model: int,
+                 nhead: int,
+                 num_encoder_layers: int,
+                 num_decoder_layers: int,
+                 dim_feedforward: int,
+                 dropout: float,
+                 activation: str,
+                 custom_encoder: nn.Module = None,
+                 custom_decoder: nn.Module = None,
+                 ):
+
+        """ PyTorch module implementing a Transformer to be used in `TransformerModel`.
+
+        PyTorch module implementing a simple encoder-decoder transformer architecture.
+
+        Parameters
+        ----------
+        input_size
+            The dimensionality of the TimeSeries instances that will be fed to the fit function.
+        output_size
+            The dimensionality of the output time series.
+        output_length
+            Number of time steps to be output by the forecasting module.
+        d_model
+            the number of expected features in the transformer encoder/decoder inputs.
+        nhead
+            the number of heads in the multiheadattention model.
+        num_encoder_layers
+            the number of encoder layers in the encoder.
+        num_decoder_layers
+            the number of decoder layers in the decoder.
+        dim_feedforward
+            the dimension of the feedforward network model.
+        dropout
+            Fraction of neurons affected by Dropout.
+        activation
+            the activation function of encoder/decoder intermediate layer, 'relu' or 'gelu'.
+        custom_encoder
+            a custom transformer encoder provided by the user (default=None)
+        custom_decoder
+            a custom transformer decoder provided by the user (default=None)
+
+        Inputs
+        ------
+        x of shape `(batch_size, input_length, input_size)`
+            Tensor containing the features of the input sequence.
+
+        Outputs
+        -------
+        y of shape `(batch_size, output_length, output_size)`
+            Tensor containing the (point) prediction at the last time step of the sequence.
+        """
+
+        super(_TransformerModule, self).__init__()
+
+        self.input_size = input_size
+        self.output_size = output_size
+        self.output_length = output_length
+
+        self.encoder = nn.Linear(input_size, d_model)
+        self.positional_encoding = PositionalEncoding(d_model, dropout)
```

Comment:
Since the _positional encoding_ class was not meant to be exposed to the user, I hid it. Also, the transformer only needs _input_length_ values out of the _max_len_ that are produced, so I initialized _max_len_ to that value